GH-73991: Rework `pathlib.Path.rmtree()` into `delete()` (#122368)
python/cpython: Doc/library/pathlib.rst

      Added return value, return the new :class:`!Path` instance.

> .. method:: Path.delete(ignore_errors=False, on_error=None)

Reviewer: At risk of bikeshedding, perhaps ``remove``? 'Delete' doesn't seem to be a frequently-used term for OS-level operations -- see PowerShell's [`Remove-Item`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.management/remove-item?view=powershell-7.4), or [``rm(tree)``](https://man7.org/linux/man-pages/man1/rm.1.html) in Unix. (DOS has ``del``, but PowerShell chose to use 'remove' over 'delete').

We have [`os.remove`](https://docs.python.org/dev/library/os.html#os.remove), so on a synonyms argument you can go both ways -- good in that this is a superset and does what I expect, or bad in that it doesn't exactly mirror ``os.remove``.

If this is part of a grand plan that I am missing though, do ignore me entirely.

A

Author: It's a good suggestion, I'll give it some thought.

My initial reaction: I reckon "delete" is more common in GUI file browsers, e.g. my Linux box offers a "Delete" option in the context menu, and macOS offers "Delete Immediately..." for items in the Bin. And I think "delete" is a tiny bit more specific, e.g. consider the difference between *removing* furniture from a room and *deleting* furniture from a room. I suspect that OSs use "remove" because of hard links: `rm some/file` may not delete the file data if its multiply hardlinked (also the reason for the `unlink()` name). Given pathlib's beginner-friendly nature it feels OK to me to gloss over the distinction. At least, that's how I feel right now :)

Author: I think I'm -1 on `remove()`.

It's true that `pathlib.Path` and `os` have methods and functions that share names, and that in some cases the pathlib methods can do more than just wrap the os functions (`mkdir()` being the most prominent example, with its *parents* and *exist_ok* arguments).

But there's no precedent for diverging as much as `os.remove()` and `pathlib.Path.remove()` would here, and diverging without requiring the user to pass additional arguments. The potential for confusion would be too great!

I do appreciate you sharing your suggestion all the same.

Author: FWIW, I've also considered `rm()`. This doesn't suffer from the `remove()` problem, but it's rather cryptic for non-users of unix shells (does it summon the Royal Marines?) and in other users it invites comparisons to the `rm` command that I'd rather avoid, as pathlib isn't in the business of precisely matching shell interfaces.

Reviewer: > does it summon the Royal Marines?

No task is too hard for the Special Boat Service! Though this would be a fun easter egg, notwithstanding the thoughts and feelings of the First Sea Lord...

> I think I'm -1 on `remove()`.

The argument you lay out is compelling, and especially given ``Path.delete``'s intended use as a one-stop-shop I think makes sense. I remember being very surprised by ``os.unlink`` when I first came across it, and ``delete`` does have the benefit of simplicity.

Thank you for entertaining my notions!

      Added return value, return the new :class:`!Path` instance.

.. method:: Path.delete(ignore_errors=False, on_error=None)

>    Delete this file or directory (including all sub-directories).

Reviewer 2: It's not possible to delete a directory without deleting sub directories (although you could raise an error).
You could maybe say "doesn't need to be empty" or "recursively deleting sub-directories"?

   If *ignore_errors* is true, errors resulting from failed deletions will be
   ignored. If *ignore_errors* is false or omitted, and a callable is given as
   the optional *on_error* argument, it will be called with one argument of
   type :exc:`OSError` each time an exception is raised. The callable can
   handle the error to continue the deletion process or re-raise it to stop.
   Note that the filename is available as the :attr:`~OSError.filename`
   attribute of the exception object. If neither *ignore_errors* nor
   *on_error* are supplied, exceptions are propagated to the caller.

   .. note::

      On platforms that lack the necessary file descriptor-based functions,
      :func:`~Path.delete` implementation is susceptible to a symlink attack:
      given proper timing and circumstances, attackers can manipulate symlinks
      on the filesystem to delete files they would not be able to access
      otherwise.

   .. versionadded:: 3.14

.. method:: Path.unlink(missing_ok=False)

>    Remove this file or symbolic link.  If the path points to a directory,

Reviewer 2: Should we also refer to delete?

      Added return value, return the new :class:`!Path` instance.

.. method:: Path.delete(ignore_errors=False, on_error=None)

>    Recursively remove this file or directory tree.

Reviewer: One can't recursively delete a file, and an attempt to make the implications mildly clearer. I haven't checked if 'directory tree' is a very widely-used term at present, but if so it might be worth adding a definition in the glossary (e.g. does a 'tree' require files, sub folders, etc etc -- simple when you're anything more than an absolute beginner, but pathlib is enticing for beginners!)
```suggestion
   Delete this file or directory (including all sub-directories).
```

>    If *ignore_errors* is true, errors resulting from failed removals will be
>    ignored. If *ignore_errors* is false or omitted, and a function is given to
>    *on_error*, it will be called each time an exception is raised. If neither
>    *ignore_errors* nor *on_error* are supplied, exceptions are propagated to
>    the caller.
> 
>    If the optional argument *on_error* is specified, it should be a callable;
>    it will be called with one argument of type :exc:`OSError`. The
>    callable can handle the error to continue the deletion process or re-raise
>    it to stop. Note that the filename is available as the
>    :attr:`~OSError.filename` attribute of the exception object.

Reviewer: This feels duplicative as we almost entirely outline *on_error* (a function taking an exception) in the first paragraph, and go on to rehash much of it in the second. However, I tried a sketch rewrite on GH's web interface and didn't get far. 

The best I had was changing the first para to say "If *ignore_errors* is false or omitted and *on_error* is given, that function will determine the error behaviour, as described below", but it's not great.

```suggestion
   If *ignore_errors* is true, errors resulting from failed deletions will be
   ignored. If *ignore_errors* is false or omitted, and a function is given to
   *on_error*, it will be called each time an exception is raised. If neither
   *ignore_errors* nor *on_error* are supplied, exceptions are propagated to
   the caller.

   If the optional argument *on_error* is specified, it should be a callable;
   it will be called with one argument of type :exc:`OSError`. The
   callable can handle the error to continue the deletion process or re-raise
   it to stop. Note that the filename is available as the
   :attr:`~OSError.filename` attribute of the exception object.
```

Author: Agreed. Better now?

FWIW I'm considering removing the *ignore_errors* and *on_error* arguments in a subsequent PR, at least for a while, because they're underdesigned and unnecessary for most use cases. So these paragraphs might not be around for long...

Reviewer: I think ``ignore_errors`` has merit in the case of "I want to try and clean up this folder, but it isn't critical", but that could be handled explicitly by the user with a try-except. The real use (if any) is for the per-error callback, but that could perhaps be handled with an `ExceptionGroup` nowadays?

Author: I agree we'll probably need to support _some_ kind of error handling customisation, but I'm not sure exactly what it looks like. I'll see if `ExceptionGroup` fits!

   .. note::

>       On platforms that support the necessary fd-based functions, a symlink
>       attack-resistant version of :meth:`~Path.delete` is used to delete
>       directory trees. On other platforms, the :func:`~Path.delete`
>       implementation is susceptible to a symlink attack: given proper timing
>       and circumstances, attackers can manipulate symlinks on the filesystem
>       to delete files they would not be able to access otherwise.

Reviewer: This will need re-wrapping if accepted. 

Can we outline what platforms are/aren't at risk? This is quite vague currently!
```suggestion
      This function can be susceptible to a 'symlink attack':
      given proper timing and circumstances, 
      attackers can manipulate symlinks on the filesystem
      to delete files they would not be able to access otherwise.
      
      On platforms that support the necessary file descriptor-based functions, 
      a symlink attack-resistant version of :meth:`~Path.delete` is used to delete
      directory trees. 
      On other platforms, the :func:`~Path.delete`
      implementation is susceptible to a this form of attack.
```

Author: > Can we outline what platforms are/aren't at risk? This is quite vague currently!

It's complicated:

```python
({os.open, os.stat, os.unlink, os.rmdir} <= os.supports_dir_fd and
 os.listdir in os.supports_fd and os.stat in os.supports_follow_symlinks))
```

`shutil` provides an [`rmtree.avoids_symlink_ataccks`](https://docs.python.org/3/library/shutil.html#shutil.rmtree.avoids_symlink_attacks) attribute. We could do the same thing here, but we'd need to `import shutil` at a module level rather than inline. Personally I think it's worthwhile - what do you reckon?

Reviewer: Ouch! Highly cryptic.

I agree that ``Path.rmtree.avoids_symlink_attacks`` is valuable to have, not least as it means that all ``shutil`` uses can be directly translated. The only imports it would add to ``pathlib`` are ``zlib`` / ``bz2`` / ``lzma``, which take 60, 1600, and 900 microseconds on my computer, and `shutil` itself, which takes 1150, for a total of ~3.7ms extra import time.

I am concerned though that ``shutil.rmtree.avoids_symlink_attacks is False`` on Windows, though not much we can do about that here!

A

Author: Thanks! I've added `Path.rmtree.avoids_symlink_attacks`

>    Remove this directory.  The directory must be empty.

Reviewer 2: Should we refer to delete here for deleting non-empty directories?
